# Introduction to Pandas Lab

Complete the following set of exercises to solidify your knowledge of Pandas fundamentals.

### 1. Import Numpy and Pandas and alias them to `np` and `pd` respectively.

In [15]:
import pandas as pd
import numpy as np

### 2. Create a Pandas Series containing the elements of the list below.

In [16]:
lst = [5.7, 75.2, 74.4, 84.0, 66.5, 66.3, 55.8, 75.7, 29.1, 43.7]

In [17]:
a = pd.Series(lst)
print(a)

0     5.7
1    75.2
2    74.4
3    84.0
4    66.5
5    66.3
6    55.8
7    75.7
8    29.1
9    43.7
dtype: float64


### 3. Use indexing to return the third value in the Series above.

*Hint: Remember that indexing begins at 0.*

In [18]:
print(a[2])

74.4


### 4. Create a Pandas DataFrame from the list of lists below. Each sublist should be represented as a row.

In [19]:
b = [[53.1, 95.0, 67.5, 35.0, 78.4],
     [61.3, 40.8, 30.8, 37.8, 87.6],
     [20.6, 73.2, 44.2, 14.6, 91.8],
     [57.4, 0.1, 96.1, 4.2, 69.5],
     [83.6, 20.5, 85.4, 22.8, 35.9],
     [49.0, 69.0, 0.1, 31.8, 89.1],
     [23.3, 40.7, 95.0, 83.8, 26.9],
     [27.6, 26.4, 53.8, 88.8, 68.5],
     [96.6, 96.4, 53.4, 72.4, 50.1],
     [73.7, 39.0, 43.2, 81.6, 34.7]]

In [20]:
df = pd.DataFrame(b)

### 5. Rename the data frame columns based on the names in the list below.

In [21]:
b = [[53.1, 95.0, 67.5, 35.0, 78.4],
     [61.3, 40.8, 30.8, 37.8, 87.6],
     [20.6, 73.2, 44.2, 14.6, 91.8],
     [57.4, 0.1, 96.1, 4.2, 69.5],
     [83.6, 20.5, 85.4, 22.8, 35.9],
     [49.0, 69.0, 0.1, 31.8, 89.1],
     [23.3, 40.7, 95.0, 83.8, 26.9],
     [27.6, 26.4, 53.8, 88.8, 68.5],
     [96.6, 96.4, 53.4, 72.4, 50.1],
     [73.7, 39.0, 43.2, 81.6, 34.7]]

In [22]:
colnames = ['Score_1', 'Score_2', 'Score_3', 'Score_4', 'Score_5']

In [28]:
df = pd.DataFrame(b, columns = colnames)
display(df)



Score_1  Score_2  Score_3  Score_4  Score_5
0     53.1     95.0     67.5     35.0     78.4
1     61.3     40.8     30.8     37.8     87.6
2     20.6     73.2     44.2     14.6     91.8
3     57.4      0.1     96.1      4.2     69.5
4     83.6     20.5     85.4     22.8     35.9
5     49.0     69.0      0.1     31.8     89.1
6     23.3     40.7     95.0     83.8     26.9
7     27.6     26.4     53.8     88.8     68.5
8     96.6     96.4     53.4     72.4     50.1
9     73.7     39.0     43.2     81.6     34.7

### 6. Create a subset of this data frame that contains only the Score 1, 3, and 5 columns.

In [32]:
subset = df[["Score_1","Score_2"]]
display(subset)

Score_1  Score_2
0     53.1     95.0
1     61.3     40.8
2     20.6     73.2
3     57.4      0.1
4     83.6     20.5
5     49.0     69.0
6     23.3     40.7
7     27.6     26.4
8     96.6     96.4
9     73.7     39.0

### 7. From the original data frame, calculate the average Score_3 value.

In [43]:
media_s3 = df["Score_3"].mean()
print(round(media_s3, 2))

56.95


### 8. From the original data frame, calculate the maximum Score_4 value.

In [44]:
df["Score_4"].max()

88.8

### 9. From the original data frame, calculate the median Score 2 value.

In [45]:
df["Score_2"].median()

40.75

### 10. Create a Pandas DataFrame from the dictionary of product orders below.

In [48]:
orders = {'Description': ['LUNCH BAG APPLE DESIGN',
  'SET OF 60 VINTAGE LEAF CAKE CASES ',
  'RIBBON REEL STRIPES DESIGN ',
  'WORLD WAR 2 GLIDERS ASSTD DESIGNS',
  'PLAYING CARDS JUBILEE UNION JACK',
  'POPCORN HOLDER',
  'BOX OF VINTAGE ALPHABET BLOCKS',
  'PARTY BUNTING',
  'JAZZ HEARTS ADDRESS BOOK',
  'SET OF 4 SANTA PLACE SETTINGS'],
 'Quantity': [1, 24, 1, 2880, 2, 7, 1, 4, 10, 48],
 'UnitPrice': [1.65, 0.55, 1.65, 0.18, 1.25, 0.85, 11.95, 4.95, 0.19, 1.25],
 'Revenue': [1.65, 13.2, 1.65, 518.4, 2.5, 5.95, 11.95, 19.8, 1.9, 60.0]}

In [52]:
df_orders = pd.DataFrame(orders)
display(df_orders)

Description  Quantity  UnitPrice  Revenue
0              LUNCH BAG APPLE DESIGN         1       1.65     1.65
1  SET OF 60 VINTAGE LEAF CAKE CASES         24       0.55    13.20
2         RIBBON REEL STRIPES DESIGN          1       1.65     1.65
3   WORLD WAR 2 GLIDERS ASSTD DESIGNS      2880       0.18   518.40
4    PLAYING CARDS JUBILEE UNION JACK         2       1.25     2.50
5                      POPCORN HOLDER         7       0.85     5.95
6      BOX OF VINTAGE ALPHABET BLOCKS         1      11.95    11.95
7                       PARTY BUNTING         4       4.95    19.80
8            JAZZ HEARTS ADDRESS BOOK        10       0.19     1.90
9       SET OF 4 SANTA PLACE SETTINGS        48       1.25    60.00

### 11. Calculate the total quantity ordered and revenue generated from these orders.

In [62]:
print(f"Total orderes quantity: {df_orders['Quantity'].sum()} \nRevenue: {df_orders['Revenue'].sum()}")

Total orderes quantity: 2978 
Revenue: 637.0


### 12. Obtain the prices of the most expensive and least expensive items ordered and print the difference.

In [88]:
print(f"El precio mas caro es: {df_orders['UnitPrice'].max()}")
display(df_orders[df_orders['UnitPrice'] == (df_orders['UnitPrice'].max())])

print(f"\nEl precio mas barato es: {df_orders['UnitPrice'].min()}")
display(df_orders[df_orders['UnitPrice'] == (df_orders['UnitPrice'].min())])


El precio mas caro es: 11.95


Description  Quantity  UnitPrice  Revenue
6  BOX OF VINTAGE ALPHABET BLOCKS         1      11.95    11.95


El precio mas barato es: 0.18


Description  Quantity  UnitPrice  Revenue
3  WORLD WAR 2 GLIDERS ASSTD DESIGNS      2880       0.18    518.4

## Let's load another dataset for more exercisesº

In [103]:
# Run this code:
admissions = pd.read_csv('../Admission_Predict.csv')
#to remove trailing spaces at the end of the column names
admissions.columns = admissions.columns.str.rstrip()

Let's evaluate the dataset by looking at the `head` function.

In [104]:
admissions.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0           1        337          118                  4  4.5  4.5  9.65   
1           2        316          104                  3  3.0  3.5  8.00   
2           3        322          110                  3  3.5  2.5  8.67   
3           4        314          103                  2  2.0  3.0  8.21   
4           5        330          115                  5  4.5  3.0  9.34   

   Research  Chance of Admit  
0         1             0.92  
1         1             0.72  
2         1             0.80  
3         0             0.65  
4         1             0.90

### 1 - Before beginning to work with this dataset and evaluating graduate admissions data, we will verify that there is no missing data in the dataset. Do this in the cell below.

In [105]:
admissions.isnull().sum()

Serial No.           0
GRE Score            0
TOEFL Score          0
University Rating    0
SOP                  0
LOR                  0
CGPA                 0
Research             0
Chance of Admit      0
dtype: int64

###  2 -  Interestingly, there is a column that uniquely identifies the applicants. This column is the serial number column. Instead of having our own index, we should make this column our index. Do this in the cell below. Keep the column in the dataframe in addition to making it an index.

In [106]:
admissions.columns

Index(['Serial No.', 'GRE Score', 'TOEFL Score', 'University Rating', 'SOP',
       'LOR', 'CGPA', 'Research', 'Chance of Admit'],
      dtype='object')

In [109]:
new_adm = admissions.set_index("Serial No.", inplace = False)

In [111]:
display(new_adm)

GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
Serial No.                                                              
1                 337          118                  4  4.5  4.5  9.65   
2                 316          104                  3  3.0  3.5  8.00   
3                 322          110                  3  3.5  2.5  8.67   
4                 314          103                  2  2.0  3.0  8.21   
5                 330          115                  5  4.5  3.0  9.34   
...               ...          ...                ...  ...  ...   ...   
381               324          110                  3  3.5  3.5  9.04   
382               325          107                  3  3.0  3.5  9.11   
383               330          116                  4  5.0  4.5  9.45   
384               312          103                  3  3.5  4.0  8.78   
385               333          117                  4  5.0  4.0  9.66   

            Research  Chance of Admit  
Serial No.                             
1                  1             0.92  
2                  1             0.72  
3                  1             0.80  
4                  0             0.65  
5                  1             0.90  
...              ...              ...  
381                1             0.82  
382                1             0.84  
383                1             0.91  
384                0             0.67  
385                1             0.95  

[385 rows x 8 columns]

Turns out that `GRE Score` and `CGPA` also uniquely identify the data. Show this in the cell below.

In [121]:
new_adm["GRE Score"].unique()

array([337, 316, 322, 314, 330, 321, 308, 302, 323, 327, 328, 307, 311,
       317, 319, 318, 303, 325, 334, 336, 298, 295, 310, 300, 338, 340,
       331, 320, 299, 304, 313, 332, 326, 329, 339, 312, 324, 309, 315,
       301, 296, 294, 306, 305, 290, 335, 333, 297, 293], dtype=int64)

In [122]:
np.array(new_adm["GRE Score"])

array([337, 316, 322, 314, 330, 321, 308, 302, 323, 327, 328, 307, 311,
       314, 317, 319, 318, 303, 325, 328, 334, 336, 322, 298, 295, 310,
       300, 327, 338, 340, 331, 320, 299, 300, 304, 307, 308, 316, 313,
       332, 326, 322, 329, 339, 321, 327, 313, 312, 334, 324, 322, 320,
       316, 298, 300, 311, 309, 307, 304, 315, 325, 325, 327, 316, 318,
       328, 332, 336, 321, 314, 314, 329, 327, 301, 296, 294, 312, 340,
       320, 322, 340, 319, 315, 317, 314, 318, 299, 298, 301, 303, 304,
       306, 331, 332, 323, 322, 312, 314, 317, 326, 316, 329, 338, 331,
       304, 305, 321, 301, 320, 311, 310, 299, 290, 296, 327, 335, 334,
       310, 308, 301, 300, 323, 319, 326, 333, 339, 303, 323, 333, 314,
       316, 326, 318, 329, 332, 331, 340, 325, 320, 315, 326, 339, 311,
       334, 332, 321, 324, 326, 312, 315, 309, 306, 297, 315, 298, 318,
       317, 329, 322, 302, 313, 293, 311, 312, 334, 322, 323, 321, 320,
       329, 319, 309, 307, 300, 305, 299, 314, 327, 317, 335, 33

In [123]:
new_adm["GRE Score"].unique() == np.array(new_adm["GRE Score"])

<ipython-input-123-8038c1460051>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  new_adm["GRE Score"].unique() == np.array(new_adm["GRE Score"])


False

No son iguales, "GRE Score" no tiene valores únicos.

In [124]:
new_adm["CGPA"].unique()

array([9.65, 8.  , 8.67, 8.21, 9.34, 8.2 , 7.9 , 8.6 , 9.  , 9.1 , 8.3 ,
       8.7 , 8.8 , 8.5 , 8.4 , 9.5 , 9.7 , 9.8 , 7.5 , 7.2 , 7.3 , 8.1 ,
       9.4 , 9.6 , 9.2 , 7.8 , 7.7 , 9.3 , 8.85, 7.4 , 7.6 , 6.8 , 8.92,
       9.02, 8.64, 9.22, 9.16, 9.64, 9.76, 9.45, 9.04, 8.9 , 8.56, 8.72,
       8.22, 7.54, 7.36, 8.02, 9.36, 8.66, 8.42, 8.28, 8.14, 7.92, 7.66,
       8.03, 7.88, 7.84, 8.96, 9.24, 8.88, 8.46, 8.12, 8.25, 8.47, 9.05,
       8.78, 9.18, 9.46, 9.38, 8.48, 8.68, 8.34, 8.45, 8.62, 7.46, 7.28,
       8.84, 9.56, 9.48, 8.36, 9.32, 8.71, 9.35, 8.65, 9.28, 8.77, 8.16,
       9.08, 9.12, 9.15, 9.44, 9.92, 9.11, 8.26, 9.43, 9.06, 8.75, 8.89,
       8.69, 7.86, 9.01, 8.97, 8.33, 8.27, 7.98, 8.04, 9.07, 9.13, 9.23,
       8.87, 8.98, 8.94, 9.53, 8.76, 8.52, 8.54, 9.91, 9.87, 7.65, 7.89,
       9.14, 9.66, 9.78, 9.42, 9.26, 8.23, 8.07, 9.31, 9.17, 9.19, 8.37,
       7.68, 8.15, 8.73, 8.83, 8.57, 9.68, 8.09, 8.79, 8.17, 7.64, 8.01,
       7.95, 8.49, 7.87, 7.97, 8.18, 8.32, 8.55, 8.

In [125]:
np.array(new_adm["CGPA"])

array([9.65, 8.  , 8.67, 8.21, 9.34, 8.2 , 7.9 , 8.  , 8.6 , 9.  , 9.1 ,
       8.  , 8.2 , 8.3 , 8.7 , 8.  , 8.8 , 8.5 , 8.4 , 9.5 , 9.7 , 9.8 ,
       8.8 , 7.5 , 7.2 , 7.3 , 8.1 , 8.3 , 9.4 , 9.6 , 9.8 , 9.2 , 8.4 ,
       7.8 , 7.5 , 7.7 , 8.  , 8.2 , 8.5 , 9.1 , 9.4 , 9.1 , 9.3 , 9.7 ,
       8.85, 8.4 , 8.3 , 7.9 , 8.  , 8.1 , 8.  , 7.7 , 7.4 , 7.6 , 6.8 ,
       8.3 , 8.1 , 8.2 , 8.2 , 8.5 , 8.7 , 8.92, 9.02, 8.64, 9.22, 9.16,
       9.64, 9.76, 9.45, 9.04, 8.9 , 8.56, 8.72, 8.22, 7.54, 7.36, 8.02,
       9.5 , 9.22, 9.36, 9.45, 8.66, 8.42, 8.28, 8.14, 7.92, 7.66, 8.03,
       7.88, 7.66, 7.84, 8.  , 8.96, 9.24, 8.88, 8.46, 8.12, 8.25, 8.47,
       9.05, 8.78, 9.18, 9.46, 9.38, 8.64, 8.48, 8.68, 8.34, 8.56, 8.45,
       9.04, 8.62, 7.46, 7.28, 8.84, 9.56, 9.48, 8.36, 8.22, 8.47, 8.66,
       9.32, 8.71, 9.1 , 9.35, 9.76, 8.65, 8.78, 9.28, 8.77, 8.16, 9.08,
       9.12, 9.15, 9.36, 9.44, 9.92, 8.96, 8.64, 8.48, 9.11, 9.8 , 8.26,
       9.43, 9.28, 9.06, 8.75, 8.89, 8.69, 8.34, 8.

In [126]:
new_adm["CGPA"].unique() == np.array(new_adm["CGPA"])


<ipython-input-126-bc9e0d020629>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  new_adm["CGPA"].unique() == np.array(new_adm["CGPA"])


False

No son iguales, CGPA no tiene valores únicos.

### 3 - In this part of the lab, we would like to test complex conditions on the entire data set at once. Let's start by finding the number of rows where the CGPA is greater than 9 and the student has performed an investigation.

#### 4 - Now return all the rows where the CGPA is greater than 9 and the SOP score is less than 3.5. Find the mean chance of admit for these applicants.

Let's create a new column based on the TOEFL Score column, where we put True if the TOEFL value is greater than 100 and False if it is less. We are going to do it with a function and an apply. 
To do this, we first create a function that receives an argument. The function will return True if the parameter entered is greater than 100, otherwise it will return False.

Now we create a new column called "Decision" and apply to the TOEFL Score column

Create a column called `decision2` in the `admissions` dataframe. Assign 1 to this column if the value of `SOP` is greater than 3 and 0 otherwise.    
HINT (use np.where)